### ISSUE TITLE
Laravel 5 AND Lumen Support

### ISSUE THREAD

**Issue by jasonlewis** [dingo/api#213]
Greetings API developer.

Okay so I know quite a few people are jumping over to L5 to use all the latest and greatest stuff that it has to offer. That's fine. You can do that. But this package isn't L5 ready any more.

This package will support the next version of Laravel. BUT, right now, I'm waiting for the dust to settle. There's a lot happening in there and I'll change one thing to fix something and before you know it another thing will change. So, for now, I'm going to iron out the remaining issues for 4.x and merge the develop branch over to master and start tagging 0.7.* releases.

I'm hoping to get the package in a somewhat stable condition. My thoughts are to then create a branch specifically for L5 as there's a number of things that will need to be changed in order for it to be compatible.

Any feedback, post it here.

Cheers.

---

**Update 2014-05-17**

**Comment by MartijnThomas** [comment]
@jasonlewis I'm not sure if you were holding back on getting this package ready due to the removed package registration (If not, you can discard this message), but this seems to be fixed (according to Taylor at Twitter): https://twitter.com/taylorotwell/status/553007842337103872.

**Comment by jasonlewis** [comment]
I'm yet to start working on any form of L5 compatibility. Will keep this issue updated when I start.

**Comment by GrahamCampbell** [comment]
@jasonlewis I'd recommend using orchestra's "config" and "support-providers" components. I'm doing to be using them in all my laravel packages. I want to push them to the standard for laravel 5 packages.

// cc @crynobone

**Comment by jasonlewis** [comment]
Will take a look, cheers @GrahamCampbell.

**Comment by joshhornby** [comment]
Any ETA on when L5 development will start? Very eager to use this in a laravel 5 app :)

Taylor has has pretty much said this is stable now and the release is set for next week.

**Comment by jasonlewis** [comment]
Soon. I'll wait for the release next week before I push anything public. There's a few things I want to tidy up first before I branch it off to make it compatible.

**Comment by kennonb** [comment]
Can't wait! I love this package. :) :+1:

**Comment by HipsterJazzbo** [comment]
@jasonlewis How big of a job are you anticipating? Like, just enough to get it going in L5, or are you expecting a bigger refactor will be necessary?

**Comment by joshhornby** [comment]
Good question @HipsterJazzbo, I am about to start a pretty big project and would love to use this package as the base for the API. Taylor tweeted that L5 will be out either tomorrow or day after.

**Comment by jasonlewis** [comment]
I'm not entirely sure to be honest. I've got a bit on my plate at the moment so haven't had a chance yet to make anyway headway. I'll be sure to update here though once I do.

**Comment by HipsterJazzbo** [comment]
Cool. I'd love it if you just chuck a `4.2.*|5.0.*` in there, so that us guys could start hacking on it. I've been doing updates and opening PRs for days.

**Comment by jasonlewis** [comment]
I'll throw up a develop branch in a minute so you guys can have a look. I warn you though. It'll be a broken mess that I doubt will even work.

**Comment by HipsterJazzbo** [comment]
It certainly won't work. I'm keen to help out though, as much as I can without getting into things that are creator decisions.

**Comment by jasonlewis** [comment]
All good. Give me a couple and I'll have it up. :smile:

**Comment by HipsterJazzbo** [comment]
Hovering over the enter key:

`$ composer require dingo/api`

**Comment by jasonlewis** [comment]
Should be able to require `dev-develop` now. I hope.

**Comment by HipsterJazzbo** [comment]
BOOM. Thanks a bunch.

**Comment by jasonlewis** [comment]
You're welcome. :smile:

**Comment by koray-mobven** [comment]
L5 has been released officially. I hope Dingo API will be released with L5 compatibility soon.

**Comment by joshhornby** [comment]
Are you open to the community helping out to port this across @jasonlewis?

**Comment by primathon** [comment]
Can't wait until this gets up and running :) been looking forward to this packing being upgraded for quite a while.

**Comment by jasonlewis** [comment]
@joshhornby I'm more then happy to have people send PRs to the `develop` branch for some L5 compatibility. I don't have the time right this very minute. Hoping to look into it more next week. Sorry guys, just been a little busy with other things.

Feel free to send some PRs though. `develop` branch should be your target.

**Comment by foxted** [comment]
I tried to execute the tests on each "component" using these dependencies:

```
    "require": {
        "php": ">=5.4.0",
        "illuminate/support": "~5.0",
        "league/fractal": "0.11.*"
    },
    "require-dev": {
        "lucadegasperi/oauth2-server-laravel": "4.0.*@dev",
        "tymon/jwt-auth": "dev-laravel-5",
        "illuminate/routing": "~5.0",
        "illuminate/events": "~5.0",
        "illuminate/auth": "~5.0",
        "illuminate/database": "~5.0",
        "illuminate/pagination": "~5.0",
        "illuminate/console": "~5.0",
        "illuminate/filesystem": "~5.0",
        "phpunit/phpunit": "~4.0",
        "mockery/mockery": "~0.9",
        "squizlabs/php_codesniffer": "~2.0"
    }
```

Only the Http & Routing folders tests don't pass, so I assume that at least the Http & Routing component of the package need to be changed to make the whole package compatible L5.

I could be entirely wrong, but I will try to work on any issue I can solve and submit a PR.

**Comment by GrahamCampbell** [comment]
fractal doesn't support L5 just yet atm

**Comment by joshhornby** [comment]
Is there a ETA for L5 on fractal? I had a look the other day and doesn't seem much going on over there @GrahamCampbell

**Comment by GrahamCampbell** [comment]
Fractal shouldn't take too long. The changes are quite small. We basically just need a new pagination adapter.

**Comment by kodeine** [comment]
Since Jason is busy, no one proposed PR for L5 compatibility yet?

**Comment by stygiansabyss** [comment]
@GrahamCampbell looks like the pagination adaptor is done now yes?  If so, I may try seeing if I can get this working on L5 and submit a PR for it.

**Comment by GrahamCampbell** [comment]
@stygiansabyss You're going to need to change `"league/fractal": "0.11.*"` to `"league/fractal": "0.12.*"` when you do. :)

**Comment by kennonb** [comment]
So I've been toying around with this for a bit. Still not really getting anywhere. I thought I had, since I got the routes to work in Artisan, but they aren't properly resolving in the browser. I think someone with more understanding of the inner-working of Laravel might be needed to finish this up. :(

https://github.com/kennonb/api/commits/laravel5

**Comment by foxted** [comment]
Same issue on my side.... https://github.com/foxted/api/tree/develop @kennonb

**Comment by HipsterJazzbo** [comment]
@kennonb @foxted What specifically is the issue you're having?

**Comment by kennonb** [comment]
Running artisan api:routes & artisan route:list both return the registered routes but none of the routes actually resolve when trying to hit them from a browser. 

Some of the DispatcherTests fail as well in reference to named / action routes. But I can't quite figure out why they're failing. Tracing the named routes shows that they are resolving but they functions to retrieve a named route are returning "null". 

I may try and record a quick look at what I'm seeing since this probably isn't making sense. :)

**Comment by foxted** [comment]
Same problem: I see routes using the artisan command but they do not work in the browser. I thought it could be related to the RouteServiceProvider now included in L5 that gets in the way, but no idea to solve this issue so far. I was thinking to move auth & rate limiting to middleware as well. Feel free to contribute on my fork if you have any idea, lets make the job easier @jasonlewis ! :palm_tree:

**Comment by HipsterJazzbo** [comment]
I wonder if we should be doing this a bit at a time? Not trying to all individually rewrite the whole package?

Maybe we should submit little changes against develop, and coordinate on here who's doing what. That way someone can convert filter's to middleware, and someone else can do some other stuff, etc.

Thoughts?

**Comment by kennonb** [comment]
:+1: I think that's definitely a good idea.

**Comment by kennonb** [comment]
I don't know enough of the inner-workings of everything to lead / setup a task list, but I'll be glad to take some tasks if someone can come up with a plan.

**Comment by kodeine** [comment]
once the tasks plan is ready , i'll be happy to take few tasks as well

**Comment by foxted** [comment]
@HipsterJazzbo Good idea :+1:

**Comment by kennonb** [comment]
Ok, so I've been trying to go through this package more and understand how everything really works. Could we break out the Route Dispatcher into a Middleware like this? I know it's not as elegant as using the ```Route::api()``` facade, but it seems to work in my limited testing so far.

https://gist.github.com/kennonb/97ef0f4ce990ad668555

So you could do a ```Route::group(['middleware' => 'api'])``` and any routes defined in that group would of course need to use the Api's ControllerTrait.

Again, it's not as elegant... it would need a good bit more work. Thoughts?

**Comment by willishq** [comment]
I like the idea of the API middleware a lot, I think it's probably the best way moving forward and best describes how the API works and should be implemented moving forward. Also I do think that using API middleware actually is elegant!

I've been looking through the code for the entire system and there are a lot of things which need changing in order to make the package even slightly stable on Laravel 5.

**Comment by jasonlewis** [comment]
I'd like the syntax of both the L4 and L5 versions to remain the same or as similar as possible. If that means changing how the L4 version works then so be it. I'd like the only differences between the two to be under the hood things. The API should be the same or, like I said, as similar as possible.

One thing that irks me with this package is how much it intrudes on Laravel itself. I want to have a play with a few different methods of registering routes to make it less intrusive, heck, it could even be less coupled to Laravel. But that's not a huge issue.

So syntax wise... I'm fine with leaving the whole `Route::api` thing if it means less impact on the Laravel router. The package is always going to have some impact but I want to make it as little as possible.

I see a couple of alternatives...

1. `API::group`, this could utilize the existing facade. Probably have the same or maybe a little less impact on the router.
2. `['before' => 'api']` or `['middleware' => 'api']` (L4 and L5 respectively). This would probably have the least impact on the router, however there is a slight difference between the L4 and L5 versions. This probably wouldn't be an issue though.

I'm also contemplating making a domain or prefix required (one or the other). My thinking is that I could then set up some sort of middleware that intercepts requests to the domain/prefix in question and prepares them accordingly. This would, in theory, reduce the overall impact the package has on the Laravel core. So... thoughts?

**Comment by foxted** [comment]
I think the 2. would be the best approach.

The problem I am facing right now is how to handle the RouteServiceProvider available in the `app/Services` folder. I really think a middleware (or filter for L4) would spare us the trouble of dealing with this Service Provider.

**Comment by willishq** [comment]
The main thing to remember is that Laravel 4 will be deprecated shortly so supporting L4 shouldn't be your primary concern.

As I said before, my preferred approach would be the `'middleware' => 'api'` approach. You could also possibly create a `Request::api()` and `Response::api()` method and do something like this in the controller:

```php
if ($request->api()) {
    return response()->api($data);
}
return view('some.view', compact('data');
```

using the middleware to determine if the request is an API request.

**Comment by willishq** [comment]
@foxted you would still need the service provider to deal with internal API calls if you need them!

**Comment by foxted** [comment]
@willishq Exactly, that is why using a middleware is the best approach.

**Comment by jasonlewis** [comment]
The service provider would still be required anyway to bootstrap everything for the package.

**Comment by jasonlewis** [comment]
Okay I'm going to knuckle down and focus on an L5 release for ya'll. <3

**Comment by willishq** [comment]
You wouldn't _need_ the service provider if you were utilising middleware
and didn't need internal API requests, though, you would need to add an
alias to the middleware though!

>

-- 
Andrew Willis
Technical Web Developer

andrew@willisilliw.com | 07842 521 806

uk.linkedin.com/in/willisilliw/

**Comment by jasonlewis** [comment]
Package config?

**Comment by kennonb** [comment]
You're the best @jasonlewis. :+1:

I too like the Middleware approach. I was toying with that direction, but just got in a little over my head as I started to see how everything really worked, and kinda gave up. :)

**Comment by sjwdavies** [comment]
This sounds great @jasonlewis and I really want to use it a large scale project i'm working on but we've started it in Laravel 5...  :-(

Any idea when a stable, or even Beta is likely to be ready?

Keep up the awesome work

**Comment by lucasmichot** [comment]
@jasonlewis :+1:

**Comment by bweston92** [comment]
Any update on this?

**Comment by dbebawy** [comment]
Would love to use this on a Laravel 5 project! Keep up the great work!

**Comment by ddedic** [comment]
+1

**Comment by skenmy** [comment]
Adding my support for an L5 implementation - would be incredibly useful to us! :+1:

**Comment by thomaspb** [comment]
+1

**Comment by coderjp** [comment]
+1

**Comment by styxit** [comment]
+1

**Comment by robbinjanssen** [comment]
+1

**Comment by trizz** [comment]
+1

**Comment by marconuccetelli** [comment]
+1

**Comment by Scriptir** [comment]
+1

**Comment by heimasi** [comment]
+1

**Comment by DarkaOnLine** [comment]
+1

**Comment by otherjohn** [comment]
Since everyone else is doing it :)
:+1:

**Comment by luoshiben** [comment]
bump++ =) Thanks, @jasonlewis! Looks like you have an eager and grateful community.

**Comment by KenZett** [comment]
+1

**Comment by ben-ids** [comment]
+1 can't wait

**Comment by kodeine** [comment]
I dont wana stay behind :p +1

**Comment by HipsterJazzbo** [comment]
So, hey guys, I think @jasonlewis has probably got the picture by now. Maybe we should stop pillaging each other's inboxes with +1s? :P

**Comment by willishq** [comment]
@HipsterJazzbo +1 to this.

**Comment by sjwdavies** [comment]
@HipsterJazzbo +1 from me...

**Comment by xAockd** [comment]
@HipsterJazzbo +1

**Comment by primathon** [comment]
If it's not on anyone's radar yet, I'd just like to mention that I'm really hoping this package will offer more accurate support to the [JSON:API specification](http://jsonapi.org/format/), at least as far as overriding collection keys (instead of the default 'data') and correct HTTP response codes (422 on a successful delete, etc). When I was using the L4 implementation, I remember having to shoehorn in a few kludges to get things behaving properly. Can't wait to see this released!

**Comment by prasanthsd** [comment]
Any news on when we can get a beta like version that actually works?

**Comment by bweston92** [comment]
@prasanthsd well considering he's not replied in 20 days I think he's got better stuff to do. How-ever I'm sure he will be open to PRs.

@primathon you can always create your own serializers for the JSON api! https://github.com/dingo/api/wiki/Transformers#advanced-configuration

**Comment by bweston92** [comment]
@primathon why on earth would you return a 422 on a successful delete? 4xx are responses that indicate something wasn't right with the request. 200 or 204 is fine for this.

**Comment by primathon** [comment]
Yup, you're absolutely correct. Definitely meant to say 204, and that's
even what I've got in my code, but apparently my brain and fingers had a
different idea when typing that out. It also looks like the json:api spec
has been updated to no longer include [non-'data' keying](
http://jsonapi.org/format/#document-structure-top-level) of resource
collections, which appears to be a [very recent update](
https://web.archive.org/web/20150201234056/http://jsonapi.org/format/#document-structure-top-level
).

**Comment by primathon** [comment]
@bweston92 Yup, you're absolutely correct. Definitely meant to say 204, and that's even what I've got in my code, but apparently my brain and fingers had a different idea when typing that out. It also looks like the json:api spec has been updated to no longer include [non-'data' keying](http://jsonapi.org/format/#document-structure-top-level) of resource collections, which appears to be a [very recent update](https://web.archive.org/web/20150201234056/http://jsonapi.org/format/#document-structure-top-level).

**Comment by bweston92** [comment]
Wow I've just had my first look into L5 this package will take some work to redo, configuration is done differently. The router has no method for `setDispatchController` and the only way I can see doing this is via middle-ware, or re-write the router.

**Comment by jasonlewis** [comment]
Yeah I'm still toying with a bunch of things. Probably going to have to rework quite a lot of it. Going to try and slim it down a lot.

**Comment by jasonlewis** [comment]
And I'm sorry for the lack of updates. I'm doing my best to get something up but I've got to jump over a few hurdles first.

**Comment by bweston92** [comment]
Yeah I attempted it but like you said it's pretty much a re-write.



—

**Comment by jasonlewis** [comment]
I've been having a crack at this the last couple hours and it seems the router in Laravel 5 can't be replaced. It's pretty much hard coded into the core at a very early stage of the app life cycle.

While I did want to avoid doing too much with the router I did need to replace it for a couple of things, so this has thrown a spanner in the works. I'm going to have to take a step back and reassess.

Let's just say that don't be holding your breath for something in the next week. And expect a lot of things to change with the L5 version.

**Comment by develpr** [comment]
Thanks @jasonlewis <3

**Comment by furey-imagination** [comment]
Nice one @jasonlewis. Happy to patiently and gratefully wait while you take the time to get it right. Cheers!

**Comment by bweston92** [comment]
@jasonlewis do you have a plan for L5? As stated above it is pretty much a re-write on the routing section, I'm wondering if you've figured something out mentally?

**Comment by kharysharpe** [comment]
@jasonlewis This is exactly what i am looking for for my next project, how is the progress for Laravel 5 coming along?. 

Based on what I have read maybe the next release should be "version 2" and create upgrade steps similar to what you need to do from Laravel 4.2 to 5.0.

Also have you considered the middleware for routing like what @kennonb suggested here - https://gist.github.com/kennonb/97ef0f4ce990ad668555 ?

Can't wait for Dingo/API for L5 - this project ROCKS!!!

**Comment by Omranic** [comment]
This package is a piece of art, thanks @jasonlewis.
**Is there any news about the updated version v2, or at least simple compatibility with Laravel 5** ?

**Comment by bweston92** [comment]
@jasonlewis would it be best for users (maybe even some Laravel 5 users) to help have a brain storming discussion on how this might be best implemented? I really want to make the switch.

**Comment by mattkenefick** [comment]
+1

**Comment by tomschlick** [comment]
+1 for brainstorming session on this. I'm willing to contribute in any way I can once the migration plan is decided upon.

**Comment by thiagof** [comment]
+1 for a roadmap

**Comment by bweston92** [comment]
Ok should we open it as a new issue? So it's easier to read and stays on topic.



—

**Comment by crynobone** [comment]
> it seems the router in Laravel 5 can't be replaced.

Is it the same case for Lumen Framework?

**Comment by jasonlewis** [comment]
@crynobone, more so in Lumen. I've been in touch with T about it all. May be able to work something out.

No good news yet though folks. I'm happy for people to brainstorm ideas and try out some things. I'll hopefully get a little time during the week once I've got some client work out of the way to see if I can get something to work.

**Comment by JavierMartinz** [comment]
+1

**Comment by GrahamCampbell** [comment]
> Is it the same case for Lumen Framework?

Lumen doesn't use Laravel's routing. It uses fast route.

**Comment by jasonlewis** [comment]
Indeed Graham. But it's all tied directly into the application.

**Comment by nathanmarks** [comment]
As someone who's new to Laravel and currently using L5 for an API, what are the reasons why the router presents an issue?

**Comment by thiagof** [comment]
If the router can't be replaced only a fork would keep some compatibility..
right?

A package rewrite can't redesign the functionalities for L5?

2015-04-19 20:14 GMT-03:00 nathanmarks <notifications@github.com>:

> As someone who's new to Laravel and currently using L5 for an API, what
> are the reasons why the router presents an issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dingo/api/issues/213#issuecomment-94322212>.
>



-- 
*Thiago Fernandes*
Analista de Processos
34 9176- 4055 | thiago@internetbudi.com.br

**Comment by jakecarlson** [comment]
+1 for Lumen compatibility, but I assume this is somewhat out of your hands as at least some dependencies must also be Lumen-compatible (e.g., Oauth2 server).

**Comment by AndyWendt** [comment]
@nathanmarks I believe the router needs to be replaced because that is how he is adding his extra Router functionality.  See [the Dingo Router class](https://github.com/dingo/api/blob/master/src/Routing/Router.php) and the [Laravel one](https://github.com/illuminate/routing/blob/master/Router.php)

**Comment by nathanmarks** [comment]
@AndyWendt Thanks! I've actually used a combination of middleware and a controller abstract along with [fractal](https://github.com/thephpleague/fractal) to create my API. Could a similar approach be used here rather than trying to add functionality to the router itself?

**Comment by AndyWendt** [comment]
@nathanmarks To version the L5 apis that I am working on currently, I am getting the version from the accept header (using helper classes of course) and using an `if` statement inside the `routes` file to control which controllers are loaded. 

You can use a route group inside the `if` statements to apply middlewares etc. 

 I have an ApiFacade (actual facade pattern not laravel facade) object that then loads a specific VersionFactory (which provides the functionality for the facade) based on the requested version.

**Comment by kodeine** [comment]
@nathanmarks, you mind sharing what you've done with your api?

**Comment by bweston92** [comment]
Some internal requests functionality was handled by router as well I believe for returning "raw" responses.



—

**Comment by nathanmarks** [comment]
@kodeine Unfortunately I can't release the source for the app right (work stuff), but if you have some specific questions I can definitely answer them. Perhaps at some point I'll do a write up as others using L5 for an API may find it useful.

**Comment by twoSeats** [comment]
There's an example of an api in lumen using Fractal here for anyone interested:
https://github.com/laravelista/Lissandra

**Comment by vinkla** [comment]
+1 for getting Laravel 5 support done.

**Comment by sahilsarpal** [comment]
+1 for getting l5 support @jasonlewis

**Comment by jasonlewis** [comment]
I know, I know.

**Comment by nathanmarks** [comment]
@jasonlewis If you're able to open some specific issues pertaining to functionality that needs working on for L5, I'm sure some of us will be able to help contribute. Have you decided yet on the direction you're going to take?

**Comment by BardiaAfshin** [comment]
+1 throwing in my plus one. What's the ETA on this?

**Comment by SimonErich** [comment]
@jasonlewis I know you here this a lot, but what if we worked on a "simpler" version of dingo, without the routing (maybe using middleware and traits) to build similar functionality and if possible introduce automatic route handling later as an additional option.
dingo is going to change, I think that is for sure now, but maybe we can slim it down first and only get the most important things working.
I'm sure you thought about that already, but just wanted to dump my 2 cents, that I, as a user of your amazing api, wouldn't mind bigger changes, as they seem to be necessary.

**Comment by jasonlewis** [comment]
There isn't really a simpler version without routing. It needs routing.

Don't stress I've got a few things in the pipeline but I'm waiting to hear
back from Taylor about a couple things. I'm not going to say an ETA but it
shouldn't be long now.

**Comment by stygiansabyss** [comment]
Might try popping into LaraChat slack.  He is frequently on line in there.

**Comment by jasonlewis** [comment]
No he's aware. He'll get to it when he can. All is well. :-)

**Comment by twoSeats** [comment]
@jasonlewis your patience with all the "banging on the door" is inspiring. Kudos. Thanks for the updates.

**Comment by jasonlewis** [comment]
Discussion moving to #398.

**Comment by sahilsarpal15** [comment]
cool

**Comment by sahilsarpal15** [comment]
@sahilsarpal (;
